### PR TITLE
Backend: Adds route for well-known change password URL

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
This PR adds support for the [well-known](https://en.wikipedia.org/wiki/List_of_/.well-known/_services_offered_by_webservers) `/.well-known/change-password` URL, which browsers can use to point a user to the change password page if, for example, their password is compromised.

**Special notes for your reviewer**:
Previously unsupported routes were handled in `api.go` with the following:
```
r.Get("/*", reqSignedIn, hs.Index)
```

meaning that the not found handler established in `http_server.go` was never used, and consequently unsupported routes returned with HTTP status 200 rather than 404. Support for the well-known change-password URL relies on the web server returning status 404 for unknown routes, so I've change the route handling logic slightly in service of this.

More info on this topic: https://web.dev/change-password-url/

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -26,6 +26,7 @@ func (hs *HTTPServer) registerRoutes() {
 	r := hs.RouteRegister
 
 	// not logged in views
+	r.Get("/", reqSignedIn, hs.Index)
```
Comment (aknuds1) [inline pkg/api/api.go:29]: Doesn't `reqSignedIn` require the client to be signed in? That contradicts "not logged in views", right? Why is this change necessary?

File: pkg/api/http_server.go [hunk 2]
```diff
@@ -288,7 +288,7 @@ func (hs *HTTPServer) applyRoutes() {
 	// then custom app proxy routes
 	hs.initAppPluginRoutes(hs.macaron)
 	// lastly not found route
-	hs.macaron.NotFound(hs.NotFoundHandler)
+	hs.macaron.NotFound(middleware.ReqSignedIn, hs.NotFoundHandler)
```
Comment (aknuds1) [inline pkg/api/http_server.go:291]: Why is this change necessary?

File: pkg/api/user.go [hunk 3]
```diff
@@ -250,6 +250,11 @@ func ChangeUserPassword(c *models.ReqContext, cmd models.ChangeUserPasswordComma
 	return Success("User password changed")
 }
 
+// GET /.well-known/change-password
```
Comment (aknuds1) [inline pkg/api/user.go:253]: Declaration doc comments should be sentences describing the thing in question:
```suggestion
// RedirectToChangePassword handles GET /.well-known/change-password.
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Requesting changes since I'm unsure about some of the modifications.

File: pkg/api/user.go (hunk 3 above)
Comment (kaydelaney) [reply pkg/api/user.go:253]: Was following the pattern established by other functions in this file

Comment (aknuds1) [reply pkg/api/user.go:253]: Yeah, there's a lot of technical debt in the Grafana code base :) When we get proper doc comment linting enabled (it's a bit tricky) it'll catch this.

File: pkg/api/http_server.go (hunk 2 above)
Comment (kaydelaney) [reply pkg/api/http_server.go:291]: Some time in the past it was discovered that there was a security issue allowing users to bypass the login screen to a degree by pointing the browser to an unsupported route. 
https://github.com/grafana/grafana/pull/19118

File: pkg/api/api.go (hunk 1 above)
Comment (kaydelaney) [reply pkg/api/api.go:29]: Was reverting it back to a previous state (see https://github.com/grafana/grafana/blob/7b7b95341e73d0ac173e9ccef25cc129048b8701/pkg/api/api.go#L27) but I agree that it's a bit confusing

Comment (marefr) [reply pkg/api/api.go:29]: Seems like you're reverting #19118. Maybe need to verify with Dominik this change is intended/works as expected.

**Comment by aknuds1** [comment]
I'll have to get back to this, will be AFK for a while.

File: pkg/api/api.go (hunk 1 above)
Comment (aknuds1) [reply pkg/api/api.go:29]: Is the idea that #19118 can be reverted, so long as you make the change on line 291 of pkg/api/http_server.go?

@dprokop What do you think?

**Comment by aknuds1** [comment]
@kaydelaney Do you know why e2e tests are failing?

**Comment by aknuds1** [comment]
Also, has the motivation for this PR been stated, apart from the "well-known" protocol's existence?

File: pkg/api/api.go (hunk 1 above)
Comment (torkelo) [reply pkg/api/api.go:29]: I think this looks fine as the not found handler now has req signed in check

**Review by torkelo — APPROVED** [review]
Looks good to me

File: pkg/api/api.go (hunk 1 above)
Comment (dprokop) [reply pkg/api/api.go:29]: Yeah, think this is ok to revert that change. Let me double-check if we are not introducing the same bug again.

Comment (aknuds1) [reply pkg/api/api.go:29]: If we are reverting the change, can we move the route handler (`r.Get("/", reqSignedIn, hs.Index)`) outside the "not logged in views" section? It doesn't sound right to me.

Comment (dprokop) [reply pkg/api/api.go:29]: Checked - all good!

**Review by dprokop — APPROVED** [review]

**Review by aknuds1 — APPROVED** [review]
LGTM given Torkel and Dominik's comments, but please see my suggestions regarding un-exporting the route handler, since it doesn't need to be exported (every exported symbol makes maintenance harder), and that e2e tests are failing.

And I think we should move the index route handler registration outside of the "not logged in views" section, since it isn't right.

**Review by marefr — APPROVED** [review]
LGTM, but agree with Arve's comments
